backend/jottacloud/jottacloud.go

```diff
@@ -0,0 +1,799 @@
+//
+// This should return ErrDirNotFound if the directory isn't
+// found.
+func (f *Fs) List(dir string) (entries fs.DirEntries, err error) {
+	// TODO: Split into ListAll?
+	//fmt.Printf("List: %s\n", dir)
+	opts := rest.Opts{
+		Method: "GET",
+		Path:   f.filePath(dir),
+	}
+
+	var resp *http.Response
+	var result api.JottaFolder
+	err = f.pacer.Call(func() (bool, error) {
+		resp, err = f.srv.CallXML(&opts, nil, &result)
+		return shouldRetry(resp, err)
+	})
+
+	if err != nil {
+		if apiErr, ok := err.(*api.Error); ok {
+			// does not exist
+			if apiErr.StatusCode == http.StatusNotFound {
+				return nil, fs.ErrorDirNotFound
+			}
+		}
+		return nil, errors.Wrap(err, "couldn't list files")
+	}
+
+	if result.Deleted {
+		return nil, fs.ErrorDirNotFound
+	}
+
+	for i := range result.Folders {
+		item := &result.Folders[i]
+		if item.Deleted {
+			continue
+		}
+		remote := path.Join(dir, restoreReservedChars(item.Name))
+		d := fs.NewDir(remote, time.Time(item.ModifiedAt))
+		entries = append(entries, d)
+	}
+
+	for i := range result.Files {
+		item := &result.Files[i]
+		if item.Deleted || item.State != "COMPLETED" {
+			continue
+		}
+		remote := path.Join(dir, restoreReservedChars(item.Name))
+		o, err := f.newObjectWithInfo(remote, item)
+		if err != nil {
+			continue
+		}
+		entries = append(entries, o)
+	}
+	//fmt.Printf("Entries: %+v\n", entries)
+	return entries, nil
+}
+
+// Creates from the parameters passed in a half finished Object which
+// must have setMetaData called on it
+//
+// Used to create new objects
+func (f *Fs) createObject(remote string, modTime time.Time, size int64) (o *Object) {
+	// Temporary Object under construction
+	o = &Object{
+		fs:      f,
+		remote:  remote,
+		size:    size,
+		modTime: modTime,
+	}
+	return o
+}
+
+// Put the object
+//
+// Copy the reader in to the new object which is returned
+//
+// The new object may have been created if an error is returned
+func (f *Fs) Put(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	o := f.createObject(src.Remote(), src.ModTime(), src.Size())
+	return o, o.Update(in, src, options...)
+}
+
+// Mkdir creates the container if it doesn't exist
+func (f *Fs) Mkdir(dir string) error {
+	_, err := f.CreateDir(dir)
+	return err
+}
+
+// purgeCheck removes the root directory, if check is set then it
+// refuses to do so if it has anything in
+func (f *Fs) purgeCheck(dir string, check bool) (err error) {
+	root := path.Join(f.root, dir)
+	if root == "" {
+		return errors.New("can't purge root directory")
+	}
+
+	entries, err := f.List(dir)
+	if err != nil {
+		return err
+	}
+
+	if check {
+		if len(entries) != 0 {
+			return fs.ErrorDirectoryNotEmpty
+		}
+	}
+
+	opts := rest.Opts{
+		Method:     "POST",
+		Path:       f.filePath(dir),
+		Parameters: url.Values{},
+	}
+
+	opts.Parameters.Set("dlDir", "true")
+
+	var resp *http.Response
+	err = f.pacer.Call(func() (bool, error) {
+		resp, err = f.srv.Call(&opts)
```

Comment:
If you are using Call you need to explicitly close the resp.Body otherwise you'll leak a file descriptor - see its docs.

You can do this by using by using CallXML(&opts, nil, nil) or by setting NoResponse: true in opts.